field_RANKSUFF_buffer_module.fypp

```diff
@@ -0,0 +1,532 @@
+  ${ft.type}$, INTENT(IN), POINTER :: DATA(${ft.shape}$)
+  INTEGER(KIND=JPIM), INTENT(IN) :: IDX
+  LOGICAL, INTENT(IN) :: CONTIG_FIELDS
+
+  INTEGER(KIND=JPIM) :: LBOUNDS(${ft.rank}$)
+
+  LBOUNDS=LBOUND(DATA)
+  IF(CONTIG_FIELDS)THEN
+    VIEW_PTR(${','.join('LBOUNDS(%d):'%(r+1) for r in range(ft.viewRank))}$) => DATA(${','.join(':' for _ in range(ft.viewRank-1))}$,IDX,:)
+  ELSE
+    VIEW_PTR(${','.join('LBOUNDS(%d):'%(r+1) for r in range(ft.viewRank))}$) => DATA(${','.join(':' for _ in range(ft.viewRank))}$,IDX)
+  ENDIF
+
+END SUBROUTINE ${ftn}$_GET_VIEW
+
+SUBROUTINE ${ftn}$_SET_CHILDREN_STATUS(CHILDREN, ISTATUS)
+   TYPE(${f'FIELD_{ft.rank-1}{ft.suffix}'}$_PTR), INTENT(INOUT):: CHILDREN(:)
+   INTEGER(KIND=JPIM), INTENT(IN) :: ISTATUS
+   INTEGER(KIND=JPIM) :: I
+
+   DO I=1,SIZE(CHILDREN)
+      CHILDREN(I)%PTR%ISTATUS = ISTATUS
+   ENDDO
+
+END SUBROUTINE ${ftn}$_SET_CHILDREN_STATUS
+
+SUBROUTINE ${ftn}$_CHECK_CHILDREN_STATUS(CHILDREN, ISTATUS, STAT)
+   TYPE(${f'FIELD_{ft.rank-1}{ft.suffix}'}$_PTR), INTENT(INOUT):: CHILDREN(:)
+   INTEGER(KIND=JPIM), INTENT(IN) :: ISTATUS
+   LOGICAL, INTENT(OUT) :: STAT
+   INTEGER(KIND=JPIM) :: I
+
+   STAT = .TRUE.
+   DO I=1,SIZE(CHILDREN)
+      IF(CHILDREN(I)%PTR%ISTATUS /= ISTATUS) STAT = .FALSE.
+   ENDDO
+
+END SUBROUTINE ${ftn}$_CHECK_CHILDREN_STATUS
+
+SUBROUTINE ${ftn}$_OWNER_RESIZE (SELF, UBOUNDS, LBOUNDS, PERSISTENT)
+  CLASS(${ftn}$_OWNER),         INTENT(IN) :: SELF
+  INTEGER(KIND=JPIM), OPTIONAL, INTENT(IN) :: UBOUNDS(${ft.rank}$)
+  INTEGER(KIND=JPIM), OPTIONAL, INTENT(IN) :: LBOUNDS(${ft.rank}$)
+  LOGICAL, OPTIONAL,            INTENT(IN) :: PERSISTENT
+
+  CALL ABOR1("BUFFER_OWNER CANNOT BE RESIZED")
+END SUBROUTINE ${ftn}$_OWNER_RESIZE
+
+#:for buffer_type in ['OWNER', 'WRAPPER']
+SUBROUTINE ${ftn}$_${buffer_type}$_CREATE_DEVICE_DATA(SELF)
+  CLASS(${ftn}$_${buffer_type}$) :: SELF
+  INTEGER(KIND=JPIM) :: IFIELD
+  
+  IF (.NOT. ASSOCIATED (SELF%DEVPTR)) THEN
+    CALL SELF%CREATE_DEVICE_DATA
+    IF(SELF%CONTIG_FIELDS)THEN
+       #:set ar = (',').join([':' for _ in range(0, ft.rank-1)])
+       DO IFIELD=1,SELF%NUM_CHILDREN
+          SELF%CHILDREN(IFIELD)%PTR%DEVPTR => SELF%DEVPTR(${ar}$,IFIELD)
+       ENDDO
+    ELSE
+       #:set ar = (',').join([':' for _ in range(0, ft.rank-2)])
+       DO IFIELD=1,SELF%NUM_CHILDREN
+          SELF%CHILDREN(IFIELD)%PTR%DEVPTR => SELF%DEVPTR(${ar}$,IFIELD,:)
+       ENDDO
+    ENDIF
+  ENDIF
+END SUBROUTINE ${ftn}$_${buffer_type}$_CREATE_DEVICE_DATA
+
+SUBROUTINE ${ftn}$_${buffer_type}$_FINAL(SELF)
+  CLASS(${ftn}$_${buffer_type}$) :: SELF
+  INTEGER(KIND=JPIM) :: IFIELD
+  LOGICAL :: STAT
+
+  CALL ${ftn}$_CHECK_CHILDREN_STATUS(SELF%CHILDREN, SELF%ISTATUS, STAT)
+
+  ! CHILD%DEVPTR was never allocated and is a discontiguous slice of BUFFER%DEVPTR
+  ! To avoid calling CHILD%DELETE_DEVICE, we copy back all the data here and 
+  ! nullify CHILD%DEVPTR
+  IF(STAT)THEN
+     CALL SELF%SYNC_HOST_RDWR()
```

Comment:
We can just call `SELF%SYNC_HOST_RDWR` and let that check the status and manage the pullback.